lib/srv/alpnproxy/auth/auth_proxy.go

```diff
@@ -0,0 +1,168 @@
+type AuthProxyDialerService struct {
+	reverseTunnelServer sitesGetter
+	accessPoint         authGetter
+}
+
+func (s *AuthProxyDialerService) HandleConnection(ctx context.Context, conn net.Conn, connInfo alpnproxy.ConnectionInfo) error {
+	defer conn.Close()
+	clusterName, err := getClusterName(connInfo)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	authConn, err := s.dialAuthServer(ctx, clusterName)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer authConn.Close()
+	if err := s.proxyConn(ctx, conn, authConn); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func getClusterName(info alpnproxy.ConnectionInfo) (string, error) {
+	if len(info.ALPN) == 0 {
+		return "", trace.NotFound("missing ALPN value")
+	}
+	protocol := info.ALPN[0]
+	if !strings.HasPrefix(protocol, string(alpnproxy.ProtocolAuth)) {
+		return "", trace.BadParameter("auth routing prefix not found")
+	}
+	routeToCluster := strings.TrimPrefix(protocol, string(alpnproxy.ProtocolAuth))
+	cn, err := auth.DecodeClusterName(routeToCluster)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	return cn, nil
+}
+
+func (s *AuthProxyDialerService) dialAuthServer(ctx context.Context, clusterNameFromSNI string) (net.Conn, error) {
+	clusterName, err := s.accessPoint.GetClusterName()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if clusterName.GetClusterName() == clusterNameFromSNI {
+		return s.dialLocalAuthServer(ctx)
+	}
+	if s.reverseTunnelServer != nil {
+		return s.dialRemoteAuthServer(ctx, clusterNameFromSNI)
+	}
+	return nil, trace.NotFound("auth server for %q cluster name not found", clusterNameFromSNI)
+}
+
+func (s *AuthProxyDialerService) dialLocalAuthServer(ctx context.Context) (net.Conn, error) {
+	authServers, err := s.accessPoint.GetAuthServers()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if len(authServers) == 0 {
+		return nil, trace.NotFound("empty auth servers list")
+	}
+	authServerIndex := rand.Intn(len(authServers))
+	conn, err := net.Dial("tcp", authServers[authServerIndex].GetAddr())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return conn, nil
+}
+
+func (s *AuthProxyDialerService) dialRemoteAuthServer(ctx context.Context, clusterName string) (net.Conn, error) {
+	sites, err := s.reverseTunnelServer.GetSites()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	for _, s := range sites {
+		if s.GetName() != clusterName {
+			continue
+		}
+		conn, err := s.DialAuthServer()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return conn, nil
+	}
+	return nil, trace.NotFound("cluster name %q not found", clusterName)
+}
+
+func (s *AuthProxyDialerService) proxyConn(ctx context.Context, upstreamConn, downstreamConn net.Conn) error {
+	errC := make(chan error)
+	go func() {
+		defer upstreamConn.Close()
+		defer downstreamConn.Close()
+		_, err := io.Copy(downstreamConn, upstreamConn)
+		errC <- trace.Wrap(err)
+
+	}()
+	go func() {
+		defer upstreamConn.Close()
+		defer downstreamConn.Close()
```

Comment:
Should both of these be closing the connections like this? Or should it be moved outside of the goroutines?

Reply:
Yes, both connections should be closed in goroutine and this approach seems to be a common pattern in our codebase:  https://github.com/gravitational/teleport/blob/master/lib/client/client.go#L1306